File: patterns/memento/java/Memento.java
import java.util.ArrayList;
import java.util.List;

// Memento
class EditorMemento {
    private String content;

    public EditorMemento(String content) {
        this.content = content;
    }

    public String getContent() {
        return content;
    }
}

// Originator
class TextEditor {
    private String content = "";

    public void setContent(String content) {
        this.content = content;
    }

    public String getContent() {
        return content;
    }

    public EditorMemento save() {
        return new EditorMemento(content);
    }

    public void restore(EditorMemento memento) {
        content = memento.getContent();
    }
}

// Caretaker
class History {
    private List<EditorMemento> mementos = new ArrayList<>();

    public void push(EditorMemento memento) {
        mementos.add(memento);
    }

    public EditorMemento pop() {
        if (mementos.isEmpty()) {
            return null;
        }
        return mementos.remove(mementos.size() - 1);
    }
}

// Main class demonstrating Memento pattern
public class Memento {
    public static void main(String[] args) {
        TextEditor editor = new TextEditor();
        History history = new History();

        editor.setContent("Version 1");
        history.push(editor.save());
        System.out.println("Current: " + editor.getContent());

        editor.setContent("Version 2");
        history.push(editor.save());
        System.out.println("Current: " + editor.getContent());

        editor.setContent("Version 3");
        System.out.println("Current: " + editor.getContent());

        editor.restore(history.pop());
        System.out.println("After undo: " + editor.getContent());

        editor.restore(history.pop());
        System.out.println("After undo: " + editor.getContent());
    }
}

File: patterns/memento/python/Memento.py
from typing import List

# Memento
class Memento:
    def __init__(self, content: str):
        self._content = content

    def get_content(self) -> str:
        return self._content

# Originator
class TextEditor:
    def __init__(self):
        self._content = ""

    def set_content(self, content: str):
        self._content = content

    def get_content(self) -> str:
        return self._content

    def save(self) -> Memento:
        return Memento(self._content)

    def restore(self, memento: Memento):
        self._content = memento.get_content()

# Caretaker
class History:
    def __init__(self):
        self._mementos: List[Memento] = []

    def push(self, memento: Memento):
        self._mementos.append(memento)

    def pop(self) -> Memento:
        if not self._mementos:
            return None
        return self._mementos.pop()


if __name__ == "__main__":
    editor = TextEditor()
    history = History()

    editor.set_content("Version 1")
    history.push(editor.save())
    print("Current:", editor.get_content())

    editor.set_content("Version 2")
    history.push(editor.save())
    print("Current:", editor.get_content())

    editor.set_content("Version 3")
    print("Current:", editor.get_content())

    # Undo to Version 2
    editor.restore(history.pop())
    print("After undo:", editor.get_content())

    # Undo to Version 1
    editor.restore(history.pop())
    print("After undo:", editor.get_content())
